crates/flipperzero/src/furi/log/metadata.rs

```diff
@@ -0,0 +1,535 @@
+    /// *enabled*.
+    ///
+    /// [`Level`]: super::Level
+    #[inline(always)]
+    pub fn current() -> Self {
+        match unsafe { sys::furi_log_get_level() } {
+            // Default log level is defined in `furi/core/log.c` in the FlipperZero firmware.
+            sys::FuriLogLevel_FuriLogLevelDefault => Self::INFO,
+            sys::FuriLogLevel_FuriLogLevelNone => Self::OFF,
+            sys::FuriLogLevel_FuriLogLevelError => Self::ERROR,
+            sys::FuriLogLevel_FuriLogLevelWarn => Self::WARN,
+            sys::FuriLogLevel_FuriLogLevelInfo => Self::INFO,
+            sys::FuriLogLevel_FuriLogLevelDebug => Self::DEBUG,
+            sys::FuriLogLevel_FuriLogLevelTrace => Self::TRACE,
+            #[cfg(debug_assertions)]
+            unknown => unreachable!(
+                "/!\\ `LevelFilter` representation seems to have changed! /!\\ \n\
+                This is a bug (and it's pretty bad). Please contact the `flipperzero` \
+                maintainers. Thank you and I'm sorry.\n \
+                The offending repr was: {:?}",
+                unknown,
+            ),
+            #[cfg(not(debug_assertions))]
+            _ => unsafe {
+                // Using `unreachable_unchecked` here (rather than
+                // `unreachable!()`) is necessary to ensure that rustc generates
+                // an identity conversion from integer -> discriminant, rather
+                // than generating a lookup table. We want to ensure this
+                // function is a single `bl` instruction (sometimes followed by
+                // a `subs` instruction to handle `FuriLogLevelDefault`) if at
+                // all possible, because it is called *every* time a logging
+                // callsite is hit; and it is (potentially) the only code in the
+                // hottest path for skipping a majority of callsites when level
+                // filtering is in use.
+                //
+                // safety: This branch is only truly unreachable if we guarantee
+                // that no values other than the possible enum discriminants
+                // will *ever* be present. The log filter is initialized by the
+                // Flipper Zero SDK to `FuriLogLevelDefault`, which is not a
+                // valid `LevelFilter` discriminant but is specifically handled
+                // above. It is set either internally by the Flipper Zero, or
+                // through the Flipper Zero SDK. The latter we expose via the
+                // `set_max` function, which takes a `LevelFilter` parameter;
+                // this restricts the inputs to `set_max` to the set of valid
+                // discriminants. Therefore, **as long as `furi_log_set_level`
+                // is only ever called by `set_max`**, this is safe.
+                core::hint::unreachable_unchecked()
+            },
+        }
+    }
+
+    pub(crate) fn set_max(LevelFilter(level): LevelFilter) {
```

Comment:
Currently this is not exposed; in `tracing-core` it was used to automatically set the application-global maximum level filter based on all `tracing` macro usages in the app and its dependencies. Is it useful to make a public API?

I'll admit to being a bit confused as to why `furi_log_set_level` is in the SDK at all, enabling individual apps to alter the Flipper Zero-global log level rather than requiring it be configured in the settings. I guess so individual apps can temporarily override it and then set it back later?

Reply:
I'm not 100% sure, but I think the current Flipper Zero firmware is built with the Memory Protection Unit (MPU) disabled, thus all threads exist in a single memory space. So applications are really just a thread in the flat memory space.

Reply:
I decided to remove this method. If we want to expose `sys::furi_log_set_level`, it should probably be as a standalone method instead to more clearly delineate FAP logging from Flipper Zero-global logging.